cmd/cloud/installation.go

```diff
@@ -423,3 +433,146 @@ var installationShowStateReport = &cobra.Command{
 		return nil
 	},
 }
+
+// WARNING: EXPERIMENTAL
+// This command runs as a client with direct store integration.
+var installationRecoveryCmd = &cobra.Command{
+	Use:   "recover",
+	Short: "recover the basic resources of a deleted installation by recreating it.",
+	RunE: func(command *cobra.Command, args []string) error {
+		command.SilenceUsage = true
+
+		installationID, _ := command.Flags().GetString("installation")
+		databaseID, _ := command.Flags().GetString("installation-database")
+
+		logger := logger.WithFields(logrus.Fields{
+			"instance":              instanceID,
+			"installation":          installationID,
+			"installation-database": databaseID,
+		})
+
+		logger.Info("Starting installation recovery from deleted state")
+
+		sqlStore, err := sqlStore(command)
+		if err != nil {
+			return errors.Wrap(err, "failed to create datastore")
+		}
+
+		installation, err := sqlStore.GetInstallation(installationID, true, false)
+		if err != nil {
+			return errors.Wrap(err, "failed to get installation")
+		}
+		if installation == nil {
+			return errors.New("installation does not exist")
+		}
+		if installation.State != model.InstallationStateDeleted {
+			return errors.New("installation recovery can only be performed on deleted installations")
+		}
+
+		// DNS could have been claimed by a new installation, so we need to check
+		// that as well.
+		installations, err := sqlStore.GetInstallations(&model.InstallationFilter{
+			DNS:    installation.DNS,
+			Paging: model.AllPagesNotDeleted(),
+		}, false, false)
+		if err != nil {
+			return errors.Wrap(err, "failed to get cluster installations filtered by DNS")
+		}
+		if len(installations) != 0 {
+			return errors.Errorf("the requested installation's DNS is now in use by %d installations", len(installations))
+		}
+
+		// Be extra cautious until we can test other configs.
+		if installation.Database != model.InstallationDatabaseMultiTenantRDSPostgres {
+			return errors.Errorf("installation database type %s is not supported", installation.Database)
+		}
+		if installation.Filestore != model.InstallationFilestoreBifrost && installation.Filestore != model.InstallationFilestoreMultiTenantAwsS3 {
+			return errors.Errorf("installation filestore type %s is not supported", installation.Filestore)
+		}
+
+		clusterInstallations, err := sqlStore.GetClusterInstallations(&model.ClusterInstallationFilter{
+			InstallationID: installation.ID,
+			Paging:         model.AllPagesWithDeleted(),
+		})
+		if err != nil {
+			return errors.Wrap(err, "failed to get cluster installations")
+		}
+		if len(clusterInstallations) != 1 {
+			return errors.Errorf("expected to find 1 cluster installation, but found %d", len(clusterInstallations))
+		}
+		clusterInstallation := clusterInstallations[0]
+
+		db, err := sqlStore.GetMultitenantDatabase(databaseID)
```

Comment:
nit: We might want to fetch database immediately before locking or even after the lock is acquired. Not sure how long is the operation of restoring secret but due to nature of `Installations` field on multitenant database, if other provisioner added entry to it, we might actually remove it on update.

Reply:
Oh very good point. I wanted to verify all the resources existed before actions were taken so let me refresh the db resource once we have it under lock.